src/Moq.Analyzers/AsShouldBeUsedOnlyForInterfaceAnalyzer.cs

```diff
@@ -29,43 +29,58 @@ public override void Initialize(AnalysisContext context)
-    private static void Analyze(SyntaxNodeAnalysisContext context)
-    {
-        if (context.Node is not InvocationExpressionSyntax invocationExpression)
+        context.RegisterCompilationStartAction(static context =>
         {
-            return;
-        }
+            // Ensure Moq is referenced in the compilation
+            ImmutableArray<INamedTypeSymbol> mockTypes = context.Compilation.GetMoqMock();
+            if (mockTypes.IsEmpty)
+            {
+                return;
+            }
 
-        if (invocationExpression.Expression is not MemberAccessExpressionSyntax memberAccessSyntax)
-        {
-            return;
-        }
+            // Look for the Mock.As() method and provide it to Analyze to avoid looking it up multiple times.
+            ImmutableArray<IMethodSymbol> asMethods = mockTypes
+                .SelectMany(mockType => mockType.GetMembers("As"))
+                .OfType<IMethodSymbol>()
+                .Where(method => method.IsGenericMethod)
+                .ToImmutableArray();
+            if (asMethods.IsEmpty)
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context => Analyze(context, asMethods), OperationKind.Invocation);
+        });
+    }
 
-        if (!MoqAsMethodDescriptor.IsMatch(context.SemanticModel, memberAccessSyntax, context.CancellationToken))
+    private static void Analyze(OperationAnalysisContext context, ImmutableArray<IMethodSymbol> wellKnownAsMethods)
+    {
+        if (context.Operation is not IInvocationOperation invocationOperation)
         {
             return;
         }
 
-        if (!memberAccessSyntax.Name.TryGetGenericArguments(out SeparatedSyntaxList<TypeSyntax> typeArguments))
+        IMethodSymbol targetMethod = invocationOperation.TargetMethod;
+        if (!targetMethod.IsInstanceOf(wellKnownAsMethods))
         {
             return;
         }
 
-        if (typeArguments.Count != 1)
+        ImmutableArray<ITypeSymbol> typeArguments = targetMethod.TypeArguments;
+        if (typeArguments.Length != 1)
         {
             return;
         }
 
-        TypeSyntax typeArgument = typeArguments[0];
-        SymbolInfo symbolInfo = context.SemanticModel.GetSymbolInfo(typeArgument, context.CancellationToken);
-
-        if (symbolInfo.Symbol is ITypeSymbol { TypeKind: not TypeKind.Interface })
+        if (typeArguments[0] is ITypeSymbol { TypeKind: not TypeKind.Interface })
         {
-            context.ReportDiagnostic(Diagnostic.Create(Rule, typeArgument.GetLocation()));
+            // Try to locate the type argument in the syntax tree to report the diagnostic at the correct location.
+            // If that fails for any reason, report the diagnostic on the operation itself.
+            NameSyntax? memberName = context.Operation.Syntax.DescendantNodes().OfType<MemberAccessExpressionSyntax>().Select(mae => mae.Name).FirstOrDefault();
```

Comment:
Note from @MattKotsenas : convert to `SingleOrDefault()`

Reply:
fixed!